### PR TITLE
Monorepo turbine into packages

### PR DESCRIPTION
**Pull request by jmar910**
Coverts our repo into a monorepo that contains _2_ publishable npm packages: `@meroxa/turbine-js-cli` && `@meroxa/turbine-js-framework`

## `turbine-js-cli`
Contains all of the logic and dependencies for initializing, testing, and deploying a turbine application. This package is what the meroxa CLI interacts with when running CLI commands on the user's computer. This package is NOT a dependency for a user's data app and does not get shipped to funtime with the user's data app. Exposes `turbine-cli` executable that accepts the standard cli commands (test, run, deploy, version)

## `turbine-js-framework`
Contains all of the function logic for booting a function service in funtime using the user's data app. Since this IS a dependency for a user's data app, this can also contain any classes, functions, utils, etc. that we'd like to give a user to use in their data apps. Exposes `turbine-js-framework` executable that boots the function server using the user's data app.

## "Build" / Dockerfile improvements
With the preceding architecture, we have dramatically reduced the amount of 'build' steps we have to do, which had a side effect of reducing our Dockerfile size by a significant amount. Since the function service logic is contained in the `turbine-framework` dependency in the user's data app, we deploy the user's data app like any standard node app with docker instead of trying to build the server into their app outside the normal `node install` channels. See https://github.com/meroxa/cli/pull/411 for more information on the reduced build logic and the removal of our temporary directory building. 

## Performance improvements
We've reduced the image size by now only shipping to funtime what we need, and not the extra CLI code + it's dependencies. Also deploy time has decreased slightly due to no more temporary directory building. 

## Versioning functionality
In order to have a robust versioning system, it makes sense to have turbine as a dependency of a user's data app (At launch this was not the case, and while it reduced dependencies, it wasn't a good enough solution for managing versions/upgrades/interface changes and also not good for giving the user importable utils). It makes more sense to have turbine as a dependency of the user's app but only ship what we use in funtime to funtime.

## General Refactors
+ Converts our function server code to TS, which was the last major piece of vanilla js in the project (a long time coming lol)
+ Breaks tests out into their related packages
+ Add changesets for managing package publishing and releases
+ Use `npm` and move away from `yarn` to standardize

## Testing
The code should be reviewed, we should agree on the publishing names, and manually publish the packages to npm. We can then test via the https://github.com/meroxa/cli/pull/411 branch with the published packages. More info coming soon....

### CODE REVIEW

File: packages/turbine-cli/src/runner/primary.ts [hunk 1]
```diff
@@ -12,27 +11,21 @@ import { Result, Ok, Err } from "ts-results";
...
-        filter: filterFunc,
-      });
-      return Ok(tmpDir);
+      await fs.copy(
+        pathToDockerfile,
+        path.join(this.pathToDataApp, "Dockerfile")
+      );
+      return Ok(true);
     } catch (e) {
-      await this.#cleanupTmpDir(tmpDir);
```
Comment (jayjayjpg) [inline packages/turbine-cli/src/runner/primary.ts:34]: Would the cleanup now be handled elsewhere? I'm not super familiar with the line of events yet

**Review by jayjayjpg — COMMENTED** [review]
First off, this already reads great and the descriptive PR documentation is awesome! I was wondering what your thoughts were on namespacing the packages further:

If, in the future, other language libraries (e.g. python) should be split out into their own sub packages, would it be helpful to distinguish the js ones with a specific module name?

I'm not the best with naming, but I imagine something along the lines of

- `@meroxa/turbine-js-cli`
- `@meroxa/turbine-js-framework`

would do. But what do you think about this approach?

**Comment by jmar910** [comment]
> First off, this already reads great and the descriptive PR documentation is awesome! I was wondering what your thoughts were on namespacing the packages further:
> 
> If, in the future, other language libraries (e.g. python) should be split out into their own sub packages, would it be helpful to distinguish the js ones with a specific module name?
> 
> I'm not the best with naming, but I imagine something along the lines of
> 
> * `@meroxa/turbine-js-cli`
> * `@meroxa/turbine-js-framework`
> 
> would do. But what do you think about this approach?

I love these suggestions. I tried to stay away from tacking `-js` onto the end of the names because NPM discourages this type of naming since its assumed all packages are JS, but I think in the middle it makes way more sense, since we want to still differentiate between the other languages.

File: packages/turbine-cli/src/runner/primary.ts (hunk 1 above)
Comment (jmar910) [reply packages/turbine-cli/src/runner/primary.ts:34]: Whoops I need to remove this extra console.log.

There is no more temporary directory building with these changes, all we do here is copy the Dockerfile from the package into the user's data app project. 

The cleanup of the dockerfile itself is handled by the meroxa CLI, similar to how it is handled for turbine-go apps.

Comment (jayjayjpg) [reply packages/turbine-cli/src/runner/primary.ts:34]: Ah sweet, and thank you for the outline!

**Review by jayjayjpg — APPROVED** [review]
This works great locally and should be good to go after another rebase! :sparkles:

**Pull request merged**